.github/workflows/e2e_on_pull.yaml

```diff
@@ -0,0 +1,164 @@
+  #
+  # Currently it will not build the podvm, instead it downloads the qcow2 file
+  # from the built image. The file will be archived so that downstream jobs can
+  # just download the file on their runners.
+  podvm:
+    name: podvm
+    needs: [authorize]
+    runs-on: ubuntu-latest
+    strategy:
+      fail-fast: true
+      matrix:
+        os:
+          - centos
+          - ubuntu
+        provider:
+          - generic
+        arch:
+          - amd64
+    env:
+      registry: quay.io/confidential-containers
+      podvm_image: podvm-${{ matrix.provider }}-${{ matrix.os }}-${{ matrix.arch }}
+      qcow2: podvm-${{ matrix.provider }}-${{ matrix.os }}-${{ matrix.arch }}.qcow2
+    steps:
+      - name: Checkout Code
+        uses: actions/checkout@v3
+
+      - name: Extract the podvm qcow2
+        run: ./hack/download-image.sh ${{ env.registry }}/${{ env.podvm_image }} . -o ${{ env.qcow2 }}
+        working-directory: podvm
+
+      - uses: actions/upload-artifact@v3
+        with:
+          name: ${{ env.qcow2 }}
+          path: podvm/${{ env.qcow2 }}
+          retention-days: 1
+
+  # Build and push the cloud-api-adaptor image
+  #
+  # By using a reusable `workflow_call` workflow we are hitting two
+  # GHA limitations here:
+  #
+  # - Cannot access the `env` context from the `with` so that it cannot
+  #   reuse the E2E_IMG_* environment variables set at this workflow level.
+  # - Cannot call a reusable workflow from a job's step, so the we cannot
+  #   merge the `image` and `prep_env` into a single one (unless we create
+  #   another reusable workflow and, well, likely hit another limitation...).
+  #
+  # Reference: https://docs.github.com/en/actions/using-workflows/reusing-workflows#limitations
+  #
+  image:
+    uses: ./.github/workflows/caa_build_and_push.yaml
+    needs: [authorize]
+    with:
+      registry: ghcr.io/${{ github.repository_owner }}
+      dev_tags: ci-pr${{ github.event.number  }}-dev
+      release_tags: ci-pr${{ github.event.number  }}
+    secrets: inherit
+
+  # Edit the kustomize files under the install directory to reference the
+  # built cloud-api-adaptor images. The entire directory is archived so that
+  # downstream jobs can simply download and use the prepared installation
+  # files.
+  #
+  # IMPORTANT: If you are enabling e2e tests for a given provider,
+  # then please update the PROVIDERS list (space-separated names, e.g.,
+  # "aws libvirt").
+  prep_install:
+    needs: [image]
+    runs-on: ubuntu-latest
+    env:
+      PROVIDERS: "libvirt"
+    steps:
+      - name: Checkout Code
+        uses: actions/checkout@v3
+
+      - name: Install kustomize
+        run: |
+          command -v kustomize >/dev/null || \
+          curl -s "https://raw.githubusercontent.com/kubernetes-sigs/kustomize/master/hack/install_kustomize.sh" | \
+            bash -s /usr/local/bin
+
+      - name: Update kustomization configuration
+        run: |
+          providers=(${{ env.PROVIDERS }})
```

Comment:
Could providers be a matrix?

Reply:
Yes, they could but I thought it would be overkill. My idea is to have a single `install` directory to be shared by all downstream jobs (i.e. provider specific jobs). If I convert to gha matrix it will create N jobs (N VMs instantiated) to produce the same artifact, it seems just resources waste.